plenum/server/rbftrequest.py

```diff
@@ -0,0 +1,358 @@
+                self.tpc_requests[inst_id].is_committed())
+
+    def has_propagate(self, sender: str) -> bool:
+        """
+        Check whether the request specified has already been propagated.
+        """
+        return sender in self.propagates
+
+    def votes(self) -> int:
+        """
+        Get the number of propagates
+        """
+        return len(self.propagates)
+
+    # EVENTS processing
+    def _propagate(self, request: Request, sender: str,
+                   quorum: Quorum, dry: bool=False):
+        """
+        Add the specified request to the list of received PROPAGATEs.
+
+        Try to finalize the request if the required conditions are met.
+
+        Determines whether to finalize client REQUESTs to replicas, based on the
+        following logic:
+
+        - If exactly quorum of PROPAGATE requests are received, then finalize.
+        - If less than quorum of requests then probably there's no consensus on the
+            REQUEST, don't finalize.
+        - If more than quorum then already finalized, so don't finalize
+
+        Even if the node hasn't received the client REQUEST itself, if it has
+        received enough number of PROPAGATE messages for the same, the REQUEST
+        can be finalized.
+
+        :param request: the REQUEST to add
+        :param sender: the name of the node sending the msg
+        :param quorum: quorum for PROPAGATES
+        """
+        if request.key != self.request.key:
+            raise ValueError(
+                "{} expects requests with key {} but {} was passed"
+                .format(self, self.request.key, request.key))
+        elif self.has_propagate(sender):
+            raise ValueError(
+                "{} Propagate from sender {} was alredy registered"
+                .format(self, sender))
+
+        _sender = self.propagates.get(sender)
+        try:
+            self.propagates[sender] = request
+
+            reason = None
+
+            if self.finalised is None:
+                # If not enough Propogates, don't bother comparing
+                if not quorum.is_reached(self.votes()):
+                    reason = 'not enough propagates'
+                else:
+                    senders = self._senders_for_request_with_quorum(quorum)
+
+                    if senders:
+                        logger.debug("{} finalizing request".format(self))
+                        # use arbitrary request as they should be the same
+                        self._finalize(senders.pop(), dry)
+                    else:
+                        reason = 'not enough the same propagates'
+
+            if reason:
+                logger.debug("{} not finalizing since {}".format(
+                    self, reason))
+        finally:
+            if dry:
+                if _sender is None:
+                    del self.propagates[sender]
+                else:
+                    self.propagates[sender] = _sender
+
+    def _on(self, ev, dry=False):
```

Comment:
In order to make it better with the Actors, I would implement the following base workflow:
- State machine (SM) receives a message
- SM analyzes whether it can process the message in the current state
- SM processes the message
- SM goes to the new state if needed
- SM performs some actions after moving to the new state.

I think the main difference with the current implementation is that we don't analyse whether we can apply the message (event) before trying to apply it.
Probably it's OK to change in the scope of catch-up refactoring (after a talk with @spivachuk).